playbooks/roles/web/templates/etc/varnish/default.vcl.j2

```diff
@@ -33,54 +33,110 @@ acl purge {
 # sub vcl_recv #
 ################
 sub vcl_hash {
-  hash_data(req.http.host);
-  hash_data(req.url);
+  # Hash protocol and URL.
   hash_data(req.http.X-Forwarded-Proto);
-  return(hash);
+  hash_data(req.url);
+
+  # Hash host or server IP.
+  if (req.http.host) {
+      hash_data(req.http.host);
+  } else {
+      hash_data(server.ip);
+  }
+
+  # Hash cookies for requests that have them.
+  if (req.http.Cookie) {
+      hash_data(req.http.Cookie);
+  }
```

Comment:
If a site is caching requests with cookies (which is actually the default behaviour) we have to include them in the hash too.